### ISSUE TITLE
Handle increased limit of attached GCE PDs

### ISSUE THREAD

**Issue by saad-ali** [kubernetes/kubernetes#24317]
The kubernetes scheduler [prevents scheduling](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/algorithm/predicates/predicates.go#L233) of [more than 16 ](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go#L36) Google Compute Engine (GCE)/Google Container Engine (GKE)  PersistentDisks (PDs) to the same node (a round about way of trying to prevent too many pods requiring GCE PDs from getting scheduled on to the same node).

Problems with this:
- GCE now supports a variable number of PDs attached to a node ([up to 128 disks](https://cloud.google.com/compute/docs/disks/#increased_persistent_disk_limits) for machines with 8 or more cores).
- When a pod referencing a GCE PD is deleted the scheduler immediately discounts it even though the volume may still be attached to the node.
- The same limit is applied to all nodes in the cluster, i.e. it assumes the cluster is homogeneous.

Work around:
- Set the [`KUBE_MAX_PD_VOLS`](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go#L40) environment variable on the master with the desired value.

We need to consider if we want to remove the `NewMaxPDVolumeCountPredicate` from the scheduler all together? If not, how do we support different limits based on the type of nodes?

CC @kubernetes/sig-storage

**Comment by wattsteve** [comment]
Agreed. @jeremyeder in Red Hat performance engineering brought up a similar request as he's keen to explore the boundaries and behaviors when increasing the number of attached disks to a cloud instance.

**Comment by jsafrane** [comment]
I  think that some form of `NewMaxPDVolumeCountPredicate` is necessary to prevent scheduler from assigning all pods that have volumes to a single node. We want some logic that scatters them among more nodes. I agree, current `NewMaxPDVolumeCountPredicate` is imperfect and additional check should be in kubelet.

The right way would be probably to add a field to Node.Spec about max. nr. of volumes supported by the node (per volume plugin???), enforced both by the scheduler and by kubelet (or attach controller)

**Comment by thockin** [comment]
@davidopp re: opaque resource scheduling.

Jan,  I think the right answer is an opaque counted-resource type.  We must learn to query nodes or cloud-providers to learn how many slots are available and we must learn that a volume or persistent-volume which resolves to a PD counts as "1 pd attachment" for scheduling feasibility.

Obviously we don't want to hardcode PD gunk, so I describe it as an opaque resource.  Nodes offer N "cloud.google.com/persistent-disk-attach" resources (opaque to scheduler), and volume plugins augment a pod's resource request to denote how many "cloud.google.com/persistent-disk-attach" resources are needed for that pod.

**Comment by davidopp** [comment]
Agree with @thockin , the solution here is the not-yet-implemented "opaque integer resource."

I'll add some comments to #19082, which was confusingly named and I've re-titled.

**Comment by davidopp** [comment]
@saad-ali Can you say more about "When a pod referencing a GCE PD is deleted the scheduler immediately discounts it even though the volume may still be attached to the node." ?

**Comment by saad-ali** [comment]
> @saad-ali Can you say more about "When a pod referencing a GCE PD is deleted the scheduler immediately discounts it even though the volume may still be attached to the node." ?

When a pod is deleted, as far as the scheduler is concerned so are its resources (i.e. volumes). But the reality is that it takes some time after the controller is deleted for the volume to actually be detached from the node (basically the scheduler is inferring resource count in a roundabout way, existence of a pod object rather than the actual number of disks attached to a node, which means it can become inaccurate sometimes, allowing overcommitting of the node.

**Comment by davidopp** [comment]
Thanks for the clarification. That's pretty unfortunate. Which component(s) know(s) when the PD has been successfully detached? If Kubelet, then we could have Kubelet report capacity and available (rather than just capacity), but this obviously makes things more complicated.

**Comment by saad-ali** [comment]
> Which component(s) know(s) when the PD has been successfully detached? If Kubelet, then we could have Kubelet report capacity and available (rather than just capacity), but this obviously makes things more complicated.

Yes at the moment kubelet should know. Once the attach/detach controller is complete, the new controller will. But the real source of truth would be the storage providers, which could be directly queried by any component to ask "is volume X attached to node A?", everything else is just inferring (i.e. I did an attach operation that succeeded, if so it should be in this state).

**Comment by davidopp** [comment]
For the first version of however we handle this, is it OK if the scheduler assigns the pod to the node even if insufficient PDs are available, and then the Kubelet waits to start the container until sufficient PDs are available? (On the assumption that scheduler would only do this when a pod was recently terminated, so the PD will be released shortly.) This would mean some containers would have to wait to start, when nodes are close to their max number of PDs in use.

**Comment by saad-ali** [comment]
> For the first version of however we handle this, is it OK if the scheduler assigns the pod to the node even if insufficient PDs are available, and then the Kubelet waits to start the container until sufficient PDs are available? (On the assumption that scheduler would only do this when a pod was recently terminated, so the PD will be released shortly.) This would mean some containers would have to wait to start, when nodes are close to their max number of PDs in use.

Yep, I think that would fine. The current code retries if attach fails (which it would if the node was overcommitted), and if all the attach retries fail, the pod creation loop retries the whole thing. Basically the pod sits in pending state until attach succeeds (not sure if there is an eventual timeout).

**Comment by fabioy** [comment]
It seems #20262 would solve this much more elegantly (especially with heterogeneous clusters), but till then, is the recommendation to just set the KUBE_MAX_PD_VOLS env var to the new max? I fear this may introduce more "flakiness" in the cluster with the potentially longer pod scheduling times.

**Comment by saad-ali** [comment]
Yes, the work around at the moment is to set the `KUBE_MAX_PD_VOLS` env variable--the scheduler will then use that as the limit when scheduling.

The controller proposed in #20262 can't really solve this problem (at least not alone), since this is a node resource scheduling problem (node X only supports Y GCE PDs, there are Z GCE PDs already scheduled to that node...). To "do this right" the scheduler must be involved and have enough information to make a decision when scheduling pods with volumes.

**Comment by davidopp** [comment]
Yeah I think #20262 is somewhat orthogonal. We can do the opaque resource thing to have the scheduler account for heterogenous max number of PDs, independent of whether we do #20262.

**Comment by fabioy** [comment]
Fair enough -- doesn't really matter who queries/tracks the max PD's per node, just that someone is doing it for the scheduler (and the controller) to do the "right thing".

**Comment by paurullan** [comment]
I stumbled with this issue and seem to be stuck since we use the hosted master in GKE instead of deploying the cluster ourselves in GCE.

I have asked this around but seems not to getting any answers.
http://serverfault.com/questions/794897/change-kube-max-pd-vols-env-var-in-master-gke

What would it be a workaround for hosted kubernetes?

**Comment by thockin** [comment]
I don't think we have a great workaround at the moment.  Can you use more machines, but smaller with less PDs?

**Comment by paurullan** [comment]
@thockin thank you for your time!

We are already using a very small machine ( n1-standard-2, 2cpu+7.5GB ) because in our experience kube-proxy  and colleges consume a constant amount of memory and in even smaller images the expense is too high.

So another node will be.

**Comment by thockin** [comment]
How many PDs are you trying to attach to a 2-core VM?

**Comment by paurullan** [comment]
@thockin I found the problem when the count was on 18 (the limit is currently on 16).
With the current plan I was hoping to put 28 (still less than the 32/64 this node supports).

**Comment by thockin** [comment]
OK, yeah, you're pretty far out there :)  This is not the most common use
case :)  I wish I had a better answer for you, but we just don't have a
generic resource for this yet.

@davidopp

**Comment by yuvipanda** [comment]
We ran into this just now too - we run a disk per pod and tried using larger nodes. Has there been any progress on this? (Also on GKE)

**Comment by davidopp** [comment]
We do have the opaque counted resource type now
https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/#opaque-integer-resources-alpha-feature

I've lost track of the plan for next steps there, e.g. moving it to beta.

cc/ @ConnorDoyle

**Comment by davidopp** [comment]
(Obviously we'd need to do additional work to publish max PDs as such a resource.)

**Comment by msau42** [comment]
I had a quick discussion with @davidopp about this.  I don't think the opaque integer resources feature will work for implementing this for the following reasons:

1.  In the pod spec, you do not explicitly specify how many PDs you want to request as a field.  The count is inferred from how many direct GCE PD volumes + how many GCE PD PVs that a pod's PVCs are bound to.  For opaque integer resources, you need to request the same resource in the pod spec that the node is exporting.
2.  The opaque integer resource is specified per container, but volumes are per pod.

For those reasons, it may be better just to have a new resource type for provider attachable disks (needs a better name), and a predicate to look for that specific resource in the node status.

Or, if there's a need for a node to have max limits on multiple types of volumes, maybe some sort of feature similar to opaque integer resources, but tailored specifically for handling volumes.

**Comment by thockin** [comment]
At some point we discussed that integer-resource plugins would have an
opportunity to look at the whole pod spec, and return the number of units
they claimed.  That implies that the apiserver will make a callout to some
service, pass the pod, etc.  Kind of yuck.

It could be async.  Admission control puts an initializer on, async service
updates the pod resources with the correct count.

**Comment by davidopp** [comment]
> At some point we discussed that integer-resource plugins would have an
opportunity to look at the whole pod spec, and return the number of units
they claimed. 

Ah, interesting. I don't remember this. Do you (or maybe @ConnorDoyle) remember where this was discussed? The current implementation of opaque integer resources (on the consumption side) is just some small code in the scheduler that says treat resources that appear in the Container.Resources and whose name starts with `pod.alpha.kubernetes.io/opaque-int-resource-` as integers, and debit the request from the corresponding allocatable.

IIUC what you are envisioning is a pretty large change, requiring an admission controller to address (1) from @msau42's list, scheduler changes to implement the plugin model you described to address (2), and running a DaemonSet on every node to discover the right value (this is true of opaque integer resources today). Maybe it's worth it if we think this machinery would be generally useful, but OTOH if "max networked disk per node" is a concept that exists on all cloud providers, maybe what @msau42 described is justifiable (kubelet publishes the value, and a new scheduler predicate function does the checking).

(Nobody has ever accused me of being overly principled.)

**Comment by thockin** [comment]
How do you know which type of PV is "the cloud one"?  What about when
everything moves to flex volumes and is pretty opaque?  What if there's
more than one "cloud one"?

**Comment by msau42** [comment]
Each volume plugin that wants to enforce some sort of node max, could register two functions:
1. Running in the admission controller, given a pod spec, return resource-x: count
2. Running during node startup, return resource-x: max count

This approach can enforce count limits independently per volume type.

**Comment by thockin** [comment]
On Tue, Apr 18, 2017 at 11:08 AM, Michelle Au <notifications@github.com> wrote:
>
> Each volume plugin that wants to enforce some sort of node max, could register two functions:
>
> Running in the admission controller, given a pod spec, return resource-x: count
> Running during node startup, return resource-x: max count
>
> This approach can enforce count limits independently per volume type.

This implies running volume drivers on the master, which is true for
built-in drivers today, but probably won't be true for CSI-based
drivers, which will have to run in-cluster. :(

**Comment by msau42** [comment]
For CSI, it could be possible to have a generic pod spec -> volume count function.  Assuming the CSI volume is similar to Flex, the volume will still have the driver name, so then the resource name just has to match the driver name.

**Comment by thockin** [comment]
can you flesh that out more?  I feel like I am missing something...

**Comment by msau42** [comment]
In the flex volume spec, you need to specify the driver name, like:
``` yaml
    flexVolume:
      driver: "k8s/nfs"
      fsType: "nfs"
      options:
        server: "172.16.0.25"
        share: "dws_nas_scratch"
```

So you could have the resource name include the flex driver name, like `pod.alpha.kubernetes.io/opaque-int-resource-flex-k8s-nfs`, that the flex plugin could publish for each driver on each node.

And on the scheduler side, you could have one flex predicate that reconstructs the resource name by extracting the driver name from the flex volume spec in the pod or PV.

**Comment by dw33z1lP** [comment]
Will it be possible in the future to set a custom KUBE_MAX_PD_VOLS for GKE (yes, at your own risk), ? For exemple 32 for a 1 vCPU/node cluster.

**Comment by msau42** [comment]
We are thinking that each node automatically sets a limit based on the VM instance type, as documented here: https://cloud.google.com/compute/docs/machine-types

**Comment by dw33z1lP** [comment]
Sure, it's better. But I wanted to know if there is a workaround to increase this limit now in GKE. 

Or do you think it will come soon in k8s next releases ?

**Comment by msau42** [comment]
I believe currently the workaround is to set it on the master node, which won't work for GKE.  We don't have cycles to address this for 1.7, but we may be able to something for 1.8.

**Comment by ymsaout** [comment]
@msau42 I didn't see anything about PD in release blog post. https://cloudplatform.googleblog.com/2017/09/google-container-engine-kubernetes-18.html

**Comment by msau42** [comment]
Unfortunately, we don't have this support in GKE yet.

@ymsaout can you describe your use case?

**Comment by ymsaout** [comment]
Thank you, I just need to increase the number of PD per compute in GKE, ideally 64.

**Comment by msau42** [comment]
Is it possible to use more smaller nodes in the meantime?

**Comment by ymsaout** [comment]
I can have 16 PD with Shared-core computes ?

There are other benefits tho have big compute for us...

**Comment by msau42** [comment]
Yes, that's correct.  I just meant using smaller nodes as a workaround to save some compute costs if you would not be able to leverage the whole compute power due to the PD limit.  Do you have any reasons to still use bigger nodes right now with the PD limit?

**Comment by ymsaout** [comment]
Thank you, we may test this. Yes, eg monitoring costs (datadog, stackdriver...) and custom admin tools for computes...

**Comment by jingxu97** [comment]
Currently since we only have limits for GCE and Azure, I found one quick solution for this problem is to change PodFitsResources() https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/algorithm/predicates/predicates.go

From the table https://cloud.google.com/compute/docs/machine-types, the number is directly related to machine type which is labled in the node spec. So we can use the machine type to map to the number of PD limits. By checking the pod spec and calculating the total number of PDs requested by the pod, we can check whether pod can fit the node based on PD limits.

**Comment by khenidak** [comment]
@jdumars FYI

**Comment by klausenbusk** [comment]
> Each volume plugin that wants to enforce some sort of node max, could register two functions:
> 
>     Running in the admission controller, given a pod spec, return resource-x: count
>     Running during node startup, return resource-x: max count
> 
> This approach can enforce count limits independently per volume type.

Is there any issue tracking this? At the momemt there seems to be no way, to limit the amount of pod k8s schedule to a node with X volumeType (ex a flexvolume plugin).

Right now I use a flexplugin to use [DigitalOcean Block Storage](https://www.digitalocean.com/products/storage/) as volume (with help from a [dynamic provisioner](https://github.com/kubernetes-incubator/external-storage/tree/master/digitalocean)). DigitalOcean only support 5 disks per node, but nothing is preventing k8s from scheduling more than that to a node.

**Comment by gnufied** [comment]
FWIW - this problem applies to EBS as well. We noticed that number of volumes that can be attached to certain instance types is lower than current limit hardcoded in scheduler - https://github.com/kubernetes/kubernetes/issues/59015

There is another aspect of this problem, which we haven't tracked is - volumes attached outside k8s (or unknown to k8s). For example, if Max volume limit is 10 on AWS and for some reason, 5 volumes are attached to the node outside k8s. This can also happen when some error condition in attach/detach controller has caused volume information to be lost from attach/detach controller. 

In this case - scheduler will keep trying to schedule pods with volumes on a node which is already maxed out. The current mechanism of counting `MaxPDVolumes` via running pods fails in that case.

**Comment by gnufied** [comment]
Another way of perhaps fixing this is to - introduce volume based taints to a node. The idea is - any component(whether controller-manager or a CSI plugin) can taint a node if the node reaches certain volume limit. 

We used similar mechanism in https://github.com/kubernetes/kubernetes/pull/55558 . But problem there is - applying a unschedulable taint(in current form) to a node, stops all kind of pods from getting scheduled to a node. I was thinking, we can introduce a taint that only prevents pod with volumes from being scheduled to a node, if node is affected by the taint.

The trick even then is perhaps - the taint must distinguish between emptydir/hostpath volumes and GCE/EBS volumes and taint will apply to the later but not the former.

**Comment by andyzhangx** [comment]
Hi, is there any update on this issue?

**Comment by gnufied** [comment]
@andyzhangx We are going to fix this in 1.11 . There is a feature repo - https://github.com/kubernetes/features/issues/554

Here is a WIP design proposal for the work - https://docs.google.com/document/d/1r_5cYRclF9pPO_m_qUwuj5t1e4-gWxmiyxFanPgOnhc/edit?usp=sharing

**Comment by andyzhangx** [comment]
@gnufied awesome! Just loop me in the PR, I would love to verify it in the first beginning.

**Comment by msau42** [comment]
This is available in alpha as of 1.11.

/remove-lifecycle rotten

**Comment by msau42** [comment]
This is available in beta as of 1.12

**Comment by gnufied** [comment]
This should be available as beta feature in 1.12.
